GSoC 2025: Implement Complete OWTF Proxy System with HTTPS Interception, Live Interceptor, Repeater, and Comprehensive Documentation

Nav-Prak opened the pull request:
This pull request transforms OWTF's proxy significantly, implementing HTTPS interception, live request modification, and comprehensive request replay capabilities.

## Description
The system now provides a complete proxy solution, featuring an intuitive web interface with real-time request interception, automated traffic modification through configurable rules, and a powerful repeater tool for testing and validating security findings. The HTTPS interception system enables security professionals to inspect encrypted traffic in real-time, automatically generating and managing SSL certificates for any domain while maintaining full browser compatibility. The interceptor framework allows for automated header manipulation, content transformation, URL rewriting, and performance testing through configurable rules that execute in priority order. The live interceptor provides interactive control over requests, allowing users to manually inspect, modify, or block traffic before it reaches its destination, while the repeater tab enables comprehensive request analysis and modification for thorough security testing. The implementation includes a modern, responsive web interface built with Bootstrap, comprehensive REST API endpoints for automation and integration, extensive documentation covering setup, configuration, and troubleshooting, and a robust testing framework to ensure reliability.

## Motivation and Context
This change is done as part of the Google Summer of Code 2025 as part of modernizing the OWTF proxy

## Reviewers
@viyatb @7a 

## How Has This Been Tested?
**Testing Environment:**
- Docker container running OWTF backend on Ubuntu 22.04 LTS
- Python 3.11 environment with Tornado web framework
- Multiple browsers (Chrome, Firefox) for HTTPS interception validation
- Various HTTPS test sites including Google, httpbin.org, and example.com

**Test Coverage:**

- **HTTPS Interception Tests**
  - Basic SSL handshake validation with curl and browser clients
  - Certificate generation and validation for multiple domains
  - Bidirectional data forwarding between client and upstream server
  - Error handling and recovery for SSL handshake failures
  - Browser compatibility testing with Chrome and Firefox

- **Integration Tests**
  - Proxy startup and configuration validation
  - CA certificate installation and trust verification in browsers
  - Request/response logging with rate limiting and file rotation
  - Interceptor functionality across all types (header, body, URL, delay)
  - CORS configuration and cross-origin request handling

- **Frontend Tests**
  - Repeater functionality with various HTTP methods and payloads
  - Proxy history display with filtering and search capabilities
  - Statistics dashboard showing real-time metrics
  - Bootstrap UI responsiveness across different screen sizes
  - Tab navigation and component integration

- **API Tests**
  - All proxy API endpoints (history, stats, CA cert, repeater, live interceptor)
  - Request/response validation and error handling
  - Rate limiting and CORS header verification
  - Authentication and access control (where applicable)

- **Documentation Validation**
  - All code examples tested and verified functional
  - Installation procedures validated in clean environments
  - Troubleshooting steps confirmed with actual error scenarios
  - API documentation examples tested for accuracy

- **Performance Tests**
  - Connection handling under moderate load conditions
  - Memory usage monitoring during extended operation
  - Certificate generation performance for multiple domains
  - Log file rotation and cleanup mechanisms

## Screenshots
- **Proxy History**
<img width="2548" height="1241" alt="image" src="https://github.com/user-attachments/assets/e58a5c4c-92b0-479c-985d-eea594563ade" />
<img width="2530" height="1063" alt="image" src="https://github.com/user-attachments/assets/a4f7c440-eea5-4e45-9f22-8171e0cb9568" />
<img width="2500" height="1299" alt="image" src="https://github.com/user-attachments/assets/4370355d-8070-4818-a4c1-923719c817ae" />

- **Interceptor**
<img width="2559" height="1256" alt="image" src="https://github.com/user-attachments/assets/aca2ac00-691b-4cda-9020-59190d18f37a" />
<img width="2559" height="1242" alt="image" src="https://github.com/user-attachments/assets/8b62c4e5-44d4-4094-bd01-e94fe6011961" />

- **Repeater**
<img width="2557" height="1232" alt="image" src="https://github.com/user-attachments/assets/10aa9b46-2b34-438e-a166-ec146aeb6aa9" />


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style (modified PEP8) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

Nav-Prak committed:
Added package.json, removed o-saft from Dockerfile and converted script files to unix format

Nav-Prak committed:
Fixed issues in package.json and Dockerfile.frontend, updated Dockerfile.backend to latest syntax

Nav-Prak committed:
Fix proxy compatibility issues: tornado.gen.Task and HTTP 500 errors - Fixed tornado.gen.Task compatibility with newer Tornado versions - Simplified CONNECT method to resolve SSL handshake issues - HTTP proxy now working correctly (resolves 500 errors) - HTTPS tunneling working without SSL interception

Nav-Prak committed:
Update proxy.py

Implemented HTTP/S request/response logging, fixed issues with log size, buffer management, infinite loops

Nav-Prak committed:
Update socket_wrapper.py

Implemented HTTP/S request/response logging, fixed issues with log size, buffer management, infinite loops

Nav-Prak committed:
Added proxy tab to the webapp, created API for proxy history, fixed bugs

Nav-Prak committed:
Update settings.py

Updated inbound proxy address

Nav-Prak committed:
Improved Proxy tab page layout

Nav-Prak committed:
feat: implement core proxy interceptor system with header/body/URL modification and delay injection capabilities

Nav-Prak committed:
Implemented complete Proxy Interceptor system and Repeater with backend HTTP proxy

Nav-Prak committed:
Enhance proxy frontend UI/UX and API functionality

- Improve Repeater tab with raw request/response viewing, better button styling, and layout fixes
- Add inline request name editing and header validation warnings
- Implement common header presets and quick-add functionality
- Add response size display and basic request validation
- Fix duplicate request naming and remove redundant UI elements
- Remove non-functional Settings tab from proxy section
- Add CA certificate download endpoint for HTTPS proxy setup
- Enhance proxy history display and entry detail components
- Update proxy page actions, reducers, and selectors for better state management
- Improve overall proxy interface styling and user experience

Nav-Prak committed:
Implement live interceptor and enhance proxy functionality: Add live HTTP request interception, repeater tab, enhanced proxy history, and improved UI/UX. Fix CORS configuration and integrate all features into existing proxy system.

Nav-Prak committed:
Push missed files from previous commit

Nav-Prak committed:
Merge branch 'develop' of https://github.com/owtf/owtf into develop

Nav-Prak committed:
Fix problems with HTTPS and added HTTPS interception capabilities. Updated documentation and test scripts

Nav-Prak committed:
Enhance certificate generation and testing:
- Set proper validity dates for generated certificates, starting from the current time and valid for one year.
- Add Subject Alternative Names (SANs) to improve browser compatibility, including main domain, www subdomain, wildcard subdomains, and localhost variations.
- Introduce unit tests for certificate generation to ensure file creation and validity checks.

Nav-Prak committed:
Implement enhanced HTTPS interception and testing framework:
- Introduce a comprehensive test script for HTTPS interception, validating request/response handling through the proxy.
- Add detailed documentation for HTTPS interception, including setup, configuration, and troubleshooting.
- Improve certificate generation logic to ensure proper handling of Subject Alternative Names (SANs) and validity periods.
- Enhance logging capabilities for better monitoring of intercepted traffic and proxy operations.

viyatb commented:
@Nav-Prak can you add some screenshots for the UI changes? thank you!

viyatb commented on owtf/api/handlers/base.py:
```diff
@@ -53,11 +53,18 @@ def set_default_headers(self):
+        if origin and origin in ALLOWED_ORIGINS:
+            self.set_header("Access-Control-Allow-Origin", origin)
+        else:
+            # Fallback to allow localhost development
```
we should check the settings/config file to see if this is in dev or production mode before using the catch all header

Nav-Prak replied:
Updated to check DEBUG variable to check if in dev or production mode

viyatb commented on owtf/api/handlers/proxy.py:
```diff
@@ -281,3 +254,516 @@ def delete(self):
+                url_pattern = data.get("url_pattern")
+                methods = data.get("methods")
+
+                from owtf.proxy.live_interceptor import LiveInterceptor
```
can you move the imports from middle of the code to top of the file?

Nav-Prak replied:
Moved to the top

viyatb commented on owtf/api/handlers/proxy.py:
```diff
@@ -281,3 +254,516 @@ def delete(self):
+            elif action == "disable":
+                # Disable live interception
+                from owtf.proxy.live_interceptor import LiveInterceptor
+
```
ditto

Nav-Prak replied:
Moved all the imports to the top

viyatb commented on owtf/api/handlers/proxy.py:
```diff
@@ -281,3 +254,516 @@ def delete(self):
+                    self.write({"error": "Missing request_id or decision"})
+                    return
+
+                from owtf.proxy.live_interceptor import LiveInterceptor
```
here too

viyatb commented on owtf/proxy/proxy.py:
```diff
@@ -347,6 +353,54 @@ def get(self):
+
+                    # Wait for user decision (with timeout)
+                    start_time = time.time()
+                    while time.time() - start_time < 30:  # 30 second timeout
```
the 30 sec constant should be defined at the top with a variable.

Nav-Prak replied:
Updated to be a constant

viyatb commented on owtf/proxy/proxy.py:
```diff
@@ -347,6 +353,54 @@ def get(self):
+                            # Clean up
+                            self.application.live_interceptor.cleanup_request(request_id)
+                            break
+                        time.sleep(0.1)  # Small delay to avoid busy waiting
```
same here for 0.1 delay

Nav-Prak replied:
Updated to be a constant

viyatb commented on owtf/proxy/test_https_interception.py:
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python3
```
can you put the tests in the tests directory? and make sure that it runs correctly with the testsuite

viyatb commented on owtf/tests/test_gen_cert.py:
```diff
@@ -0,0 +1,71 @@
+    d = tmp_path_factory.mktemp("ca")
+    ca_key = d / "ca.key.pem"
+    ca_crt = d / "ca.crt.pem"
+    # generate CA key and cert
```
put these in the global tests directory so it runs with the test suite.

Nav-Prak replied:
Moved all tests to the global test directory

viyatb commented on owtf/tests/test_ssl_interception.py:
```diff
@@ -0,0 +1,129 @@
+"""
+import os
+import tempfile
+import unittest
```
same with this test.

viyatb commented on owtf/tests/test_ssl_interception.py:
```diff
@@ -0,0 +1,129 @@
+        ca_cert.get_subject().OU = "Test"
+        ca_cert.get_subject().CN = "Test CA"
+        ca_cert.gmtime_adj_notBefore(0)
+        ca_cert.gmtime_adj_notAfter(365 * 24 * 60 * 60)
```
constant here - define it explicitly

Nav-Prak replied:
Defined as constant

Nav-Prak committed:
- Consolidate LiveInterceptor import statements in proxy handler for clarity.
- Move hand-coded values to constants in the proxy
- Move the tests to the global tests folder to allow integration with the test suite

Nav-Prak committed:
Update CORS handling in BaseRequestHandler to allow localhost during development when DEBUG is enabled

Nav-Prak commented:
@viyatb Thanks for reviewing! I have added the screenshots for the UI changes and made the changes as necessary

viyatb committed:
Merge branch 'develop' into develop
